input/include/NovelRT/Input/KeyStateFrameChangeLog.hpp

```diff
@@ -0,0 +1,53 @@
+// Copyright © Matt Jones and Contributors. Licensed under the MIT Licence (MIT). See LICENCE.md in the repository root
+// for more information.
+
+#ifndef NOVELRT_INPUT_KEYSTATEFRAMECHANGELOG_H
+#define NOVELRT_INPUT_KEYSTATEFRAMECHANGELOG_H
+
+#ifndef NOVELRT_INPUT_H
+#error NovelRT does not support including types explicitly by default. Please include Input.h instead for the Input namespace subset.
+#endif
+
+namespace NovelRT::Input
+{
+    class KeyStateFrameChangeLog
+    {
+
+    private:
+        KeyState _currentState = KeyState::Idle;
+        uint32_t _changeCount = 0;
+
+    public:
+        inline void PushNewState(KeyState newState) noexcept
+        {
+            if (_currentState == newState)
+            {
+                return;
+            }
+
+            _currentState = newState;
+            _changeCount++;
+        }
+
+        inline KeyState GetCurrentState() const noexcept
+        {
+            return _currentState;
+        }
+
+        inline uint32_t GetChangeCount() const noexcept
+        {
+            return _changeCount;
+        }
```

Comment:
We should avoid inlining functions as per our new header include policies. Are we updating the include structure for this in a separate submission? I don't remember.